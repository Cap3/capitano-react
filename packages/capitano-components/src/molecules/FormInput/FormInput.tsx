import * as React from "react";
import { styled, spacingAbsolute } from "@cap3/capitano-theme";
import { formElementHeight } from "../../atoms/basics/formElementHelpers";
import { Input, EllipsisText, Icons } from "../../atoms";
import { Label } from "../../atoms/Label/Label";
import {
  ValidationMessage,
  ValidationMessageProps,
} from "../../atoms/ValidationMessage/ValidationMessage";
import { InputProps } from "../../atoms/Input/Input";
import { css } from "emotion";

export type LabelPositionProp = {
  labelPosition?: "top" | "left";
};

type ValidatedProp = { validated?: boolean };
export type FormInputProps = ValidatedProp &
  InputProps &
  ValidationMessageProps &
  LabelPositionProp & {
    labelProps?: React.HtmlHTMLAttributes<HTMLLabelElement>;
    inputRef?: JSX.IntrinsicElements["input"]["ref"];
    label?: string;
    validationIcon?: boolean;
    validationMessage?: string;
  };

const iconClass = css({
  width: spacingAbsolute.spx16,
  height: spacingAbsolute.spx16,
  position: "absolute",
  right: spacingAbsolute.spx8,
  boxSizing: "border-box",
  gridArea: "input",
  color: "currentcolor",
});

const StyledLabel = styled(Label)<
  { hasLabel: boolean } & ValidatedProp &
    LabelPositionProp &
    ValidationMessageProps
>(({ labelPosition, invalid, theme, validated, hasLabel }) => {
  const leftLabel = labelPosition === "left";
  const gridTemplateAreas = leftLabel
    ? '"label input" ". message"'
    : '"label" "input" "message"';

  const validationColor =
    validated &&
    (invalid ? theme.colors.danger["500"] : theme.colors.success["500"]);

  return {
    color: validationColor || undefined,
    height: "auto",
    textAlign: leftLabel ? "right" : "left",
    display: "grid",
    gridTemplateAreas,
    gridColumnGap: hasLabel ? spacingAbsolute.spx8 : 0,
    gridTemplateColumns: leftLabel ? `${hasLabel ? 76 : 0}px 1fr` : "1fr",
    alignItems: "center",
    gridTemplateRows: leftLabel
      ? `${formElementHeight("medium", theme)}px 14px`
      : `18px ${formElementHeight("medium", theme)}px 14px`,
  };
});

const StyledInput = styled(Input)({
  gridArea: "input",
  color: "currentcolor",
});

const LabelText = styled(EllipsisText)(({ theme }) => ({
  gridArea: "label",
  color: theme.colors.layout.M60,
}));

const StyledValidationMessage = styled(ValidationMessage)({
  gridArea: "message",
  textAlign: "right",
});

export const FormInput: React.SFC<FormInputProps> = ({
  invalid,
  validationMessage,
  label,
  labelPosition = "left",
  labelProps,
  inputRef,
  validationIcon,
  validated,
  ...props
}) => {
  const validDefined = invalid !== undefined;
  const showValidationMessage = validDefined && validationMessage !== undefined;

  const showValidationIcon = validationIcon && validDefined;

  const ValidationIcon = invalid ? Icons.Close : Icons.Check;

  return (
    <StyledLabel
      hasLabel={!!label}
      invalid={invalid}
      validated={validated}
      labelPosition={labelPosition}
      {...labelProps}
    >
      {label !== undefined && <LabelText>{label}</LabelText>}
      <StyledInput {...props} innerRef={inputRef} />
      {showValidationIcon && <ValidationIcon className={iconClass} />}
      {showValidationMessage && (
        <StyledValidationMessage invalid={invalid}>
          {validationMessage}
        </StyledValidationMessage>
      )}
    </StyledLabel>
  );
};
