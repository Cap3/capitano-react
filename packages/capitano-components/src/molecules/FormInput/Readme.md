**Simple Input**

```js
const { Grid } = require("../../atoms");

<Grid columnCount={4} gap={20}>
  <FormInput />
  <FormInput label="First name" />
  <FormInput
    label="Last name"
    invalid={false}
    validationMessage="name available"
  />
  <FormInput label="Email" invalid validationMessage="invalid email adress" />
  <FormInput label="first name" labelPosition="top" />
  <FormInput
    label="Adress"
    invalid={false}
    validationMessage="adress found"
    labelPosition="top"
  />
  <FormInput
    placeholder="placeholder"
    validationIcon
    validated
    label="First name"
    labelPosition="top"
    invalid={false}
  />
  <FormInput
    placeholder="placeholder"
    label="First name"
    invalid={false}
    validationIcon
    validated
    defaultValue="First name"
  />
  <FormInput
    placeholder="placeholder"
    label="First name"
    invalid
    validationIcon
    validated
    defaultValue="First name"
  />
  <FormInput
    placeholder="placeholder"
    validationIcon
    validated
    invalid
    defaultValue="First name"
  />
</Grid>;
```

**Outlined Input**

```js
const { Grid } = require("../../atoms");

<Grid columnCount={4} gap={20}>
  <FormInput outlined={true} />
  <FormInput label="First name" outlined={true} />
  <FormInput
    label="Last name"
    invalid={false}
    validationMessage="name available"
    outlined={true}
  />
  <FormInput
    label="Email"
    invalid
    validationMessage="invalid email adress"
    outlined={true}
  />
  <FormInput label="First name" labelPosition="top" outlined={true} />
  <FormInput
    label="Adress"
    invalid={false}
    validationMessage="adress found"
    labelPosition="top"
    outlined={true}
  />
  <FormInput
    placeholder="placeholder"
    label="First name"
    labelPosition="top"
    invalid={false}
    outlined={true}
    validationIcon
    validated
  />
  <FormInput
    placeholder="placeholder"
    label="First name"
    invalid={false}
    validationIcon
    validated
    defaultValue="First name"
    outlined={true}
  />
  <FormInput
    placeholder="placeholder"
    label="First name"
    invalid
    defaultValue="First name"
    outlined={true}
    validationIcon
    validated
  />
  <FormInput
    placeholder="placeholder"
    invalid
    defaultValue="First name"
    outlined={true}
    validationIcon
    validated
  />
</Grid>;
```
