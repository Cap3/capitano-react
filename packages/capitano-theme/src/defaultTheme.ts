import { CapitanoTheme } from "./index";

// tslint:disable:max-file-line-count
export const defaultLightTheme: CapitanoTheme = {
  spacing: {
    baseUnit: 8,
    borderRadiusFactor: {
      sharp: 0,
      small: 0.5,
      medium: 1,
    },
    formElementFactor: {
      large: 6,
      medium: 4,
      small: 3,
    },
  },
  colors: {
    layout: {
      M0: "#000000",
      M10: "#333333",
      M20: "#5a5a5a",
      M30: "#787878",
      M40: "#949494",
      M50: "#b2b2b2",
      M60: "#cecece",
      M70: "#e6e6e6",
      M80: "#f7f7f7",
      M90: "#fcfcfc",
      base: "#fefefe",
    },
    textOnBackground: "#2b2b2b",
    textOnBackgroundInverse: "#fafafa",
    primary: {
      "50": "#e0eef9",
      "100": "#b3d5f1",
      "200": "#80b9e8",
      "300": "#4d9ddf",
      "400": "#2688d8",
      "500": "#0073d1",
      "600": "#006bcc",
      "700": "#0060c6",
      "800": "#0056c0",
      "900": "#0043b5",
      A100: "#dee8ff",
      A200: "#abc4ff",
      A400: "#78a0ff",
      A700: "#5e8eff",
    },
    textOnPrimary: "#fafafa",
    secondary: {
      "50": "#f6fae0",
      "100": "#e8f2b3",
      "200": "#d9ea80",
      "300": "#cae14d",
      "400": "#beda26",
      "500": "#b3d400",
      "600": "#accf00",
      "700": "#a3c900",
      "800": "#9ac300",
      "900": "#8bb900",
      A100: "#f7ffe2",
      A200: "#e8ffaf",
      A400: "#d9ff7c",
      A700: "#d2ff63",
    },
    textOnSecondary: "#fafafa",
    success: {
      "50": "#e1f8e9",
      "100": "#b5eec9",
      "200": "#84e3a5",
      "300": "#52d881",
      "400": "#2dcf66",
      "500": "#08c74b",
      "600": "#07c144",
      "700": "#06ba3b",
      "800": "#04b333",
      "900": "#02a623",
      A100: "#d0ffd7",
      A200: "#9dffab",
      A400: "#6aff7f",
      A700: "#51ff69",
    },
    textOnSuccess: "#fafafa",
    warning: {
      "50": "#fff9ed",
      "100": "#fff1d1",
      "200": "#ffe8b3",
      "300": "#ffdf94",
      "400": "#ffd87d",
      "500": "#ffd166",
      "600": "#ffcc5e",
      "700": "#ffc653",
      "800": "#ffc049",
      "900": "#ffb538",
      A100: "#ffffff",
      A200: "#ffffff",
      A400: "#fff3e1",
      A700: "#ffe9c8",
    },
    danger: {
      "50": "#fde3e9",
      "100": "#fab8c8",
      "200": "#f789a3",
      "300": "#f45a7e",
      "400": "#f13663",
      "500": "#ef1347",
      "600": "#ed1140",
      "700": "#eb0e37",
      "800": "#e80b2f",
      "900": "#e40620",
      A100: "#ffffff",
      A200: "#ffd9dc",
      A400: "#ffa6ad",
      A700: "#ff8c96",
    },
    textOnDanger: "#fafafa",
    disabled: {
      M0: "#000000",
      M10: "#111415",
      M20: "#23272a",
      M30: "#353b40",
      M40: "#475056",
      M50: "#5a656d",
      M60: "#6e7c86",
      M70: "#86939b",
      M80: "#9fa8b0",
      M90: "#b2bac0",
      base: "#bbc2c7",
    },
    textOnDisabled: "#fafafa",
  },
  statusEffects: {
    hover: {
      effect: "brighten",
      factor: 15,
    },
    pressed: {
      effect: "shade",
      factor: 15,
    },
    focus: {
      effect: "tint",
      factor: 20,
    },
  },
  typography: {
    fontFamily: {
      base: "Open Sans, sans-serif",
      monospace:
        '"Dank Mono", "Fira Code", Consolas, "Liberation Mono", Menlo, monospace',
    },
    fontSize: {
      medium: 14,
      large: 16,
      small: 12,
      h1: 32,
      h2: 28,
      h3: 24,
      h4: 20,
      h5: 16,
      h6: 14,
    },
  },
};

export const defaultDarkTheme: CapitanoTheme = {
  spacing: {
    baseUnit: 8,
    borderRadiusFactor: {
      sharp: 0,
      small: 0.5,
      medium: 1,
    },
    formElementFactor: {
      large: 6,
      medium: 4,
      small: 3,
    },
  },
  colors: {
    layout: {
      M0: "#ffffff",
      M10: "#f6f6f6",
      M20: "#eeeeee",
      M30: "#dcdcdc",
      M40: "#c8c8c8",
      M50: "#b4b4b4",
      M60: "#9d9d9d",
      M70: "#858585",
      M80: "#686868",
      M90: "#424242",
      base: "#252525",
    },
    textOnBackground: "#dedede",
    textOnBackgroundInverse: "#404040",
    primary: {
      "50": "#e0eef9",
      "100": "#b3d5f1",
      "200": "#80b9e8",
      "300": "#4d9ddf",
      "400": "#2688d8",
      "500": "#0073d1",
      "600": "#006bcc",
      "700": "#0060c6",
      "800": "#0056c0",
      "900": "#0043b5",
      A100: "#dee8ff",
      A200: "#abc4ff",
      A400: "#78a0ff",
      A700: "#5e8eff",
    },
    textOnPrimary: "#ffffff",
    secondary: {
      "50": "#f6fae0",
      "100": "#e8f2b3",
      "200": "#d9ea80",
      "300": "#cae14d",
      "400": "#beda26",
      "500": "#b3d400",
      "600": "#accf00",
      "700": "#a3c900",
      "800": "#9ac300",
      "900": "#8bb900",
      A100: "#f7ffe2",
      A200: "#e8ffaf",
      A400: "#d9ff7c",
      A700: "#d2ff63",
    },
    textOnSecondary: "#ffffff",
    success: {
      "50": "#e1f8e9",
      "100": "#b5eec9",
      "200": "#84e3a5",
      "300": "#52d881",
      "400": "#2dcf66",
      "500": "#08c74b",
      "600": "#07c144",
      "700": "#06ba3b",
      "800": "#04b333",
      "900": "#02a623",
      A100: "#d0ffd7",
      A200: "#9dffab",
      A400: "#6aff7f",
      A700: "#51ff69",
    },
    textOnSuccess: "#ffffff",
    warning: {
      "50": "#fff9ed",
      "100": "#fff1d1",
      "200": "#ffe8b3",
      "300": "#ffdf94",
      "400": "#ffd87d",
      "500": "#ffd166",
      "600": "#ffcc5e",
      "700": "#ffc653",
      "800": "#ffc049",
      "900": "#ffb538",
      A100: "#ffffff",
      A200: "#ffffff",
      A400: "#fff3e1",
      A700: "#ffe9c8",
    },
    danger: {
      "50": "#fde3e9",
      "100": "#fab8c8",
      "200": "#f789a3",
      "300": "#f45a7e",
      "400": "#f13663",
      "500": "#ef1347",
      "600": "#ed1140",
      "700": "#eb0e37",
      "800": "#e80b2f",
      "900": "#e40620",
      A100: "#ffffff",
      A200: "#ffd9dc",
      A400: "#ffa6ad",
      A700: "#ff8c96",
    },
    textOnDanger: "#ffffff",
    disabled: {
      M0: "#000000",
      M10: "#111415",
      M20: "#23272a",
      M30: "#353b40",
      M40: "#475056",
      M50: "#5a656d",
      M60: "#6e7c86",
      M70: "#86939b",
      M80: "#9fa8b0",
      M90: "#b2bac0",
      base: "#bbc2c7",
    },
    textOnDisabled: "#ffffff",
  },
  statusEffects: {
    hover: {
      effect: "brighten",
      factor: 15,
    },
    pressed: {
      effect: "shade",
      factor: 15,
    },
    focus: {
      effect: "tint",
      factor: 20,
    },
  },
  typography: {
    fontFamily: {
      base: "Open Sans, sans-serif",
      monospace:
        '"Dank Mono", "Fira Code", Consolas, "Liberation Mono", Menlo, monospace',
    },
    fontSize: {
      medium: 14,
      large: 16,
      small: 12,
      h1: 32,
      h2: 28,
      h3: 24,
      h4: 20,
      h5: 16,
      h6: 14,
    },
  },
};
