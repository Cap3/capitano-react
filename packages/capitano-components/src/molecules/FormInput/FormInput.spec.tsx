import * as React from "react";
import { ThemeProvider } from "emotion-theming";
import { FormInput } from "./FormInput";
import { render } from "react-testing-library";
import { defaultLightTheme } from "@cap3/capitano-theme/lib/defaultTheme";

describe("Input Component", () => {
  test("render dummy snapshot with react-testing-library", () => {
    const wrapper = render(
      <ThemeProvider theme={defaultLightTheme}>
        <FormInput
          placeholder="test"
          label="Adresse"
          invalid={false}
          validated
          validationIcon
          validationMessage="Adresse wurde gefunden."
          labelPosition="top"
          outlined={true}
        />
      </ThemeProvider>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });
});
