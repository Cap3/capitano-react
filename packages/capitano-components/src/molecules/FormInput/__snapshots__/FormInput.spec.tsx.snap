// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component render dummy snapshot with react-testing-library 1`] = `
.emotion-2 {
  width: 16px;
  height: 16px;
  position: absolute;
  right: 8px;
  box-sizing: border-box;
  grid-area: input;
  color: currentcolor;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  color: #838383;
  font-family: Open Sans,sans-serif;
  font-size: 14px;
  height: 32px;
  color: #00aea2;
  height: auto;
  text-align: left;
  display: grid;
  grid-template-areas: "label" "input" "message";
  grid-column-gap: 8px;
  grid-template-columns: 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-rows: 18px 32px 14px;
}

.emotion-0 {
  font-family: Open Sans,sans-serif;
  font-size: 14px;
  text-overflow: ellipsis;
  width: 100%;
  overflow: hidden;
  white-space: nowrap;
  display: inline-block;
  grid-area: label;
  color: #a1a1a1;
}

.emotion-1 {
  border-radius: 8px;
  height: 32px;
  width: auto;
  box-sizing: border-box;
  min-width: 65px;
  padding-left: 8px;
  padding-right: 8px;
  height: 32px;
  font-size: 14px;
  color: #000000;
  background-color: #ffffff;
  font-family: Open Sans,sans-serif;
  outline: none;
  border-color: #4d4d4d;
  border-width: 1px;
  border-style: solid;
  grid-area: input;
  color: currentcolor;
}

.emotion-1::-webkit-input-placeholder {
  opacity: 0.5;
  color: #000000;
}

.emotion-1:-moz-placeholder {
  opacity: 0.5;
  color: #000000;
}

.emotion-1::-moz-placeholder {
  opacity: 0.5;
  color: #000000;
}

.emotion-1:-ms-input-placeholder {
  opacity: 0.5;
  color: #000000;
}

.emotion-3 {
  color: #00aea2;
  font-family: Open Sans,sans-serif;
  font-size: 12px;
  line-height: 14;
  grid-area: message;
  text-align: right;
}

<label
  class="emotion-4"
>
  <span
    class="emotion-0"
    title="Adresse"
  >
    Adresse
  </span>
  <input
    class="emotion-1"
    placeholder="test"
  />
  <svg
    class="emotion-2"
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style="vertical-align: middle;"
    viewBox="0 0 40 40"
    width="1em"
  >
    <g>
      <path
        d="m15 27l17.7-17.7 2.3 2.3-20 20-9.3-9.3 2.3-2.3z"
      />
    </g>
  </svg>
  <div
    class="emotion-3"
  >
    Adresse wurde gefunden.
  </div>
</label>
`;
